File: skills/cumcm-code-writer/references/algorithm_library/01_优化与启发式/ant_colony_tsp/matlab/main.m
% 蚁群算法 ACO · 十城市 TSP（MATLAB 最小可运行模板）
here = fileparts(mfilename('fullpath'));
addpath(fullfile(here, '..', '..', '..', '_common', 'matlab'));

coords = [0 0; 10 0; 20 5; 22 18; 12 26; 0 22; -8 14; -6 4; 8 12; 16 10];
D = squareform(pdist(coords));
n = size(coords, 1);
ants = 10; iterations = 120; alpha = 1; beta = 2; rho = 0.5; q = 1;
tau = ones(n, n);
eta = 1 ./ (D + eye(n));
rng(42);
best = inf; bestTour = 1:n; history = zeros(iterations, 1);

for it = 1:iterations
    tours = zeros(ants, n); lengths = zeros(ants, 1);
    for k = 1:ants
        start = randi(n);
        tour = start; visited = false(1, n); visited(start) = true;
        while numel(tour) < n
            cur = tour(end);
            prob = (tau(cur, :) .^ alpha) .* (eta(cur, :) .^ beta);
            prob(visited) = 0;
            prob = prob / sum(prob);
            nxt = find(rand() <= cumsum(prob), 1);
            tour(end + 1) = nxt; %#ok<AGROW>
            visited(nxt) = true;
        end
        tours(k, :) = tour;
        idx = [tour tour(1)];
        lengths(k) = sum(D(sub2ind(size(D), idx(1:end-1), idx(2:end))));
    end
    tau = (1 - rho) * tau;
    for k = 1:ants
        idx = [tours(k, :) tours(k, 1)];
        for m = 1:n
            tau(idx(m), idx(m + 1)) = tau(idx(m), idx(m + 1)) + q / lengths(k);
        end
    end
    [curBest, kBest] = min(lengths);
    if curBest < best
        best = curBest; bestTour = tours(kBest, :);
    end
    history(it) = best;
end

mmlib_report('蚁群算法', sprintf('最优路径长度=%.3f', best), ...
    sprintf('巡回顺序=%s', mat2str(bestTour)), sprintf('迭代=%d', iterations));
mmlib_savetable(here, 'ant_colony_tsp', table((1:iterations)', history, 'VariableNames', {'迭代', '当前最优'}));

% 图件由 Python 生成（技能规则：交付图必须用 Python 绘制，MATLAB 仅负责计算）

File: skills/cumcm-code-writer/references/algorithm_library/01_优化与启发式/ant_colony_tsp/python/main.py
"""蚁群算法 ACO · 十城市 TSP（Python 最小可运行模板）

模型：min 巡回路径总长度；转移概率由信息素 τ^α 与启发式 1/d^β 决定；轮盘赌选择下一城市。
参数：蚂蚁 10 只，迭代 120 次，α=1，β=2，挥发率 0.5，信息素增量 Q/L。
接口：solve(coords, iterations, seed) -> (最优路径, 长度, 收敛序列)；运行：python main.py
"""
from __future__ import annotations

import sys
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "_common" / "python"))
from mmlib import demo, report, save_fig, save_table, set_style  # noqa: E402

COORDS = np.array([[0, 0], [10, 0], [20, 5], [22, 18], [12, 26],
                   [0, 22], [-8, 14], [-6, 4], [8, 12], [16, 10]], dtype=float)


def build_demo() -> dict:
    return {"coords": COORDS}


def solve(coords, ants: int = 10, iterations: int = 120, alpha: float = 1.0,
          beta: float = 2.0, rho: float = 0.5, q: float = 1.0, seed: int = 42):
    rng = np.random.default_rng(seed)
    diff = coords[:, None, :] - coords[None, :, :]
    dist = np.sqrt((diff ** 2).sum(-1))
    n = len(coords)
    tau = np.ones((n, n))
    eta = 1.0 / np.where(dist == 0, np.inf, dist)
    best_tour, best_length = None, np.inf
    history = []
    for _ in range(iterations):
        tours, lengths = [], []
        for _ in range(ants):
            start = int(rng.integers(0, n))
            tour = [start]
            visited = {start}
            while len(tour) < n:
                current = tour[-1]
                prob = np.zeros(n)
                for city in range(n):
                    if city not in visited:
                        prob[city] = tau[current, city] ** alpha * eta[current, city] ** beta
                prob = prob / prob.sum()
                nxt = int(rng.choice(n, p=prob))
                tour.append(nxt)
                visited.add(nxt)
            tour = np.array(tour)
            length = float(dist[tour, np.roll(tour, -1)].sum())
            tours.append(tour)
            lengths.append(length)
        tau *= (1 - rho)
        for tour, length in zip(tours, lengths):
            for a, b in zip(tour, np.roll(tour, -1)):
                tau[a, b] += q / length
        index = int(np.argmin(lengths))
        if lengths[index] < best_length:
            best_length, best_tour = lengths[index], tours[index].copy()
        history.append(best_length)
    return best_tour, best_length, history, dist


if __name__ == "__main__":
    data = demo(__file__, build_demo)
    tour, best, history, dist = solve(np.asarray(data["coords"], float))
    tour = np.roll(tour, -int(np.argmin(tour)))
    report("蚁群算法", 最优路径长度=round(best, 3), 巡回顺序=tour.tolist(), 迭代=len(history))
    save_table(__file__, "最优路径", {"顺序": np.arange(1, len(tour) + 1), "城市": tour})
    save_table(__file__, "收敛过程", {"迭代": np.arange(1, len(history) + 1), "当前最优": np.round(history, 4)})
    plt = set_style()
    fig, axes = plt.subplots(1, 2, figsize=(7.4, 3.2))
    path = np.append(tour, tour[0])
    axes[0].plot(np.asarray(data["coords"])[path, 0], np.asarray(data["coords"])[path, 1],
                 "-o", color="#009E73", markersize=4)
    axes[0].set_title(f"最优路径（长度 {best:.2f}）")
    axes[0].set_xlabel("x 坐标")
    axes[0].set_ylabel("y 坐标")
    axes[1].plot(history, color="#D55E00")
    axes[1].set_title("蚁群收敛过程")
    axes[1].set_xlabel("迭代次数")
    axes[1].set_ylabel("当前最优长度")
    save_fig(__file__, fig, "ant_colony_tsp")
